r2-core/src/main/java/com/linkedin/r2/filter/transport/FilterChainClient.java

```diff
@@ -94,6 +103,27 @@ public void streamRequest(StreamRequest request,
   public void shutdown(Callback<None> callback)
   {
     _client.shutdown(callback);
+
+    List<StreamFilter> streamFilters = _filters.getStreamFilters();
+    List<RestFilter> restFilters = _filters.getRestFilters();
+    List<StreamFilter> sharedStreamFilters = _sharedFilters.getStreamFilters();
+    List<RestFilter> sharedRestFilters = _sharedFilters.getRestFilters();
+
+    streamFilters.stream()
+      .filter(filter -> !sharedStreamFilters.contains(filter))
+      .filter(TimedStreamFilter.class::isInstance)
+      .map(TimedStreamFilter.class::cast)
+      .map(TimedStreamFilter::getTimingKeyList)
+      .flatMap(Collection::stream)
+        .forEach(TimingKey::unregisterKey);
+
+    restFilters.stream()
+      .filter(filter -> !sharedRestFilters.contains(filter))
+      .filter(TimedRestFilter.class::isInstance)
+      .map(TimedRestFilter.class::cast)
+      .map(TimedRestFilter::getTimingKeyList)
+      .flatMap(Collection::stream)
+        .forEach(TimingKey::unregisterKey);
```

Comment:
I think the usage of sharedStreamFilter is an overKill here. Probably we just need some kind of flag on the existing filters to identify if it is a global filter vs a local filter.

Reply:
Fixed